core/src/main/java/org/apache/calcite/rel/rules/SetOpToFilterRule.java

```diff
@@ -0,0 +1,241 @@
+    Pair<RelNode, RexNode> first = extractSourceAndCond(inputs.get(0).stripped());
+    // Map to store conditions grouped by <sourceRelNode, inputPosition>.
+    // Valid cond positions are set to null and others are tagged with
+    // input indices for map-based grouping.
+    Map<Pair<RelNode, Integer>, List<@Nullable RexNode>> sourceToConds =
+        new LinkedHashMap<>();
+
+    RelNode firstSource = first.left;
+    sourceToConds.computeIfAbsent(Pair.of(firstSource, null),
+        k -> new ArrayList<>()).add(first.right);
+
+    for (int i = 1; i < inputs.size(); i++) {
+      final RelNode input = inputs.get(i).stripped();
+      final Pair<RelNode, @Nullable RexNode> pair = extractSourceAndCond(input);
+      sourceToConds.computeIfAbsent(Pair.of(pair.left, pair.right != null ? null : i),
+          k -> new ArrayList<>()).add(pair.right);
+    }
+
+    if (sourceToConds.size() == inputs.size()) {
+      return;
+    }
+
+    int branchCount = 0;
+    for (Map.Entry<Pair<RelNode, Integer>, List<@Nullable RexNode>> entry
+        : sourceToConds.entrySet()) {
+      Pair<RelNode, Integer> left = entry.getKey();
+      List<@Nullable RexNode> conds = entry.getValue();
+      // If the condition is met (refer to extractSourceAndCond method),
+      // directly add its corresponding input to the new inputs list.
+      if (conds.size() == 1 && conds.get(0) == null) {
+        builder.push(left.left);
+        branchCount++;
+        continue;
+      }
+
+      RexNode combinedCond =
+          combineConditions(builder, conds, setOp, left.left == firstSource);
+
+      builder.push(left.left)
+          .filter(combinedCond);
+      branchCount++;
+    }
+
+    buildSetOp(builder, branchCount, setOp)
+        .distinct();
+    call.transformTo(builder.build());
+  }
+
+  private static RelBuilder buildSetOp(RelBuilder builder, int count, RelNode setOp) {
+    if (setOp instanceof Union) {
+      return builder.union(false, count);
+    } else if (setOp instanceof Intersect) {
+      return builder.intersect(false, count);
+    } else if (setOp instanceof Minus) {
+      return builder.minus(false, count);
+    }
+    // unreachable
+    throw new IllegalStateException("unreachable code");
+  }
+
+  private static Pair<RelNode, @Nullable RexNode> extractSourceAndCond(RelNode input) {
+    if (input instanceof Filter) {
+      Filter filter = (Filter) input;
+      if (!RexUtil.isDeterministic(filter.getCondition())
+          || RexUtil.SubQueryFinder.containsSubQuery(filter)) {
+        // If filter has non-deterministic expressions or subqueries,
+        // just return current input and null cond.
+        return Pair.of(input, null);
+      }
+      return Pair.of(filter.getInput().stripped(), filter.getCondition());
+    }
+    return Pair.of(input.stripped(),
+        input.getCluster().getRexBuilder().makeLiteral(true));
+  }
+
+  private static RexNode andFirstNotRest(RelBuilder builder, List<RexNode> conds) {
```

Comment:
a short comment would help, although the function is pretty clear
builds `conds.get(0) && !conds.get(1) && ... `